sst/SimengCoreWrapper.cc

```diff
@@ -0,0 +1,218 @@
+    cache_line_width  = params.find<uint64_t>("cache_line_width", "64");
+    max_addr_memory = params.find<uint64_t>("max_addr_memory", "0");
+
+    if (executable_path.length() == 0) {
+        output.fatal(CALL_INFO, 10, 0, "Simeng executable binary filepath not provided.");
+    }
+    if (max_addr_memory == 0) {
+      output.fatal(CALL_INFO, 10, 0, "Maximum address range for memory not provided");
+    }
+
+    iterations = 0;
+    vitrual_counter = 0;
+    size = 0;
+
+    mem = loadUserSubComponent<SST::Interfaces::StandardMem>("memory", ComponentInfo::SHARE_NONE, clock,
+      new StandardMem::Handler<SimengCoreWrapper>(this, &SimengCoreWrapper::handleEvent));
+
+    data_memory = std::make_unique<SimengMemInterface>(mem, cache_line_width, max_addr_memory, &output);
+
+    handlers = new SimengMemInterface::SimengMemHandlers(*data_memory, &output);
+
+    registerAsPrimaryComponent();
+    primaryComponentDoNotEndSim();
+}
+
+SimengCoreWrapper::~SimengCoreWrapper() {}
+
+void SimengCoreWrapper::setup() {
+  mem->setup();
+  output.verbose(CALL_INFO, 1, 0, "Memory setup complete\n");
+
+}
+
+void SimengCoreWrapper::handleEvent( StandardMem::Request* ev) {
+  ev->handle(handlers);
+}
+
+void SimengCoreWrapper::finish() {
+
+    output.verbose(CALL_INFO, 1, 0, "Simulation complete. Finalising stats....\n");
+
+    auto endTime = std::chrono::high_resolution_clock::now();
+    auto duration =
+      std::chrono::duration_cast<std::chrono::milliseconds>(endTime - start_time)
+          .count();
+    auto hz = iterations / (static_cast<double>(duration) / 1000.0);
+    auto khz = hz / 1000.0;
+    auto retired = core->getInstructionsRetiredCount();
+    auto mips = retired / static_cast<double>(duration) / 1000.0;
+
+  // Print stats
+    std::cout << "\n";
+    auto stats = core->getStats();
+    for (const auto& [key, value] : stats) {
+    std::cout << key << ": " << value << "\n";
+    }
+
+    std::cout << "\nFinished " << iterations << " ticks in " << duration << "ms ("
+            << std::round(khz) << " kHz, " << std::setprecision(2) << mips
+            << " MIPS)" << std::endl;
+
+    delete[] process_memory;
+}
+
+void SimengCoreWrapper::init(unsigned int phase) {
+    mem->init(phase);
+    if (phase == 0) {
+      fabricateSimengCore();
+    }
+}
+
+bool SimengCoreWrapper::clockTick(SST::Cycle_t current_cycle) {
+    // Tick the core and memory interfaces until the program has halted
+    if (!core->hasHalted() || data_memory->hasPendingRequests()) {
+    // Tick the core
+    core->tick();
+    // Update Virtual Counter Timer at correct frequency.
+    if (iterations % (uint64_t)timer_modulo == 0) {
+        vitrual_counter++;
+        core->incVCT(vitrual_counter);
+    }
+
+    // Tick memory
+    instruction_memory->tick();
+    data_memory->tick();
+
+    iterations++;
+
+    return false;
+    } else {
+        primaryComponentOKToEndSim();
+        return true;
+    }
+}
+
+void SimengCoreWrapper::fabricateSimengCore() {
```

Comment:
As this function looks very similar to what happens in `main.cc`, will it need manually updating if SimEng's `main.cc` is changed?

Reply:
Yes, I believed we've had this conversation before. I think the plan was to have a fabricate method in Core. But I wasn't tasked with this.

Reply:
OK, this may be something to come back to at a later stage (i.e. not in this PR)